text/0145-remove-unnecessary-allocator-usage.md

```diff
@@ -0,0 +1,1013 @@
+
+In addition to that, runtimes make substantial use of heap memory allocations, and each allocation needs to go through the runtime <-> host boundary twice (once for allocating and once for freeing). Moving the allocator to the runtime side would be a good idea, although it would increase the runtime size. But before the host-side allocator can be deprecated, all the host functions that use it must be updated to avoid using it.
+
+## Stakeholders
+
+Runtime developers, who will benefit from the improved performance and more deterministic behavior of the runtime code.
+
+## Explanation
+
+### New definitions
+
+#### <a name="new-def-i"></a>New Definition I: Runtime Optional Positive Integer
+
+The Runtime optional positive integer is a signed 64-bit value. Positive values in the range of [0..2³²) represent corresponding unsigned 32-bit values. The value of `-1` represents a non-existing value (an _absent_ value). All other values are invalid.
+
+#### <a name="new-def-ii"></a>New Definition II: Runtime Optional Pointer-Size
+
+The Runtime optional pointer-size has exactly the same definition as Runtime pointer-size ([Definition 216](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size)) with the value of 2⁶⁴-1 representing a non-existing value (an _absent_ value).
+
+### Changes to host functions
+
+#### ext_storage_get
+
+##### Existing prototype
+
+```wat
+(func $ext_storage_get_version_1
+    (param $key i64) (result i64))
+```
+
+##### Changes
+
+The function is considered obsolete, as it only implements a subset of functionality of `ext_storage_read` and uses host-allocated buffers. Users are encouraged to use `ext_storage_read_version_2` instead.
+
+#### ext_storage_read
+
+##### Existing prototype
+
+```wat
+(func $ext_storage_read_version_1
+    (param $key i64) (param $value_out i64) (param $offset i32) (result i64))
+```
+
+##### Changes
+
+The function was returning a SCALE-encoded `Option`-wrapped 32-bit integer representing the number of bytes left at supplied `offset`. It was using a host-allocated buffer to return it. It is changed to always return the full length of the value directly as a primitive value.
+
+##### New prototype
+
+```wat
+(func $ext_storage_read_version_2
+    (param $key i64) (param $value_out i64) (param $value_offset i32) (result i64))
+```
+
+##### Arguments
+
+* `key` is a pointer-size ([Definition 216](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size)) to the storage key being read;
+* `value_out` is a pointer-size ([Definition 216](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size)) to a buffer where the value read should be stored. The value is actually stored only if the buffer is large enough. Otherwise, the buffer contents are undefined;
```

Comment:
```suggestion
* `value_out` is a pointer-size ([Definition 216](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size)) to a buffer where the value read should be stored. The value is actually stored only if the buffer is large enough. Otherwise, the buffer contents are unchanged;
```

We cannot have an "undefined" anywhere. "Undefined" means "anything can happen" (including overwriting it with random values), which would break determinism.